## IBM Applied Data Science Capstone Project

## Week 5

In [111]:
import pandas as pd
import matplotlib.cm as cm
import matplotlib.colors as colors
import numpy as np

In [2]:
import requests

Data scraping

In [5]:
url = 'https://en.wikipedia.org/wiki/List_of_neighbourhoods_in_Mumbai'

In [6]:
wiki_url = requests.get(url)

In [7]:
wiki_data = pd.read_html(wiki_url.text)

In [8]:
wiki_data

[                Area                 Location   Latitude  Longitude
 0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
 1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
 2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
 3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
 4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
 ..               ...                      ...        ...        ...
 88             Parel             South Mumbai  18.990000  72.840000
 89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
 90       Dava Bazaar             South Mumbai  18.946882  72.831362
 91           Dharavi                   Mumbai  19.040208  72.850850
 92             Thane                   Mumbai  19.200000  72.970000
 
 [93 rows x 4 columns]]

In [11]:
wiki_data[0]

Area                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

Putting the data into the dataframe

In [19]:
df = pd.DataFrame(wiki_data[0])

In [20]:
df

Area                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

Importing geopy in order to find the co-ordinates of the different areas in Mumbai.

In [12]:
from geopy.geocoders import Nominatim

In [13]:
address = 'Mumbai, Maharashtra'

geolocator = Nominatim(user_agent="mumbai_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Mumbai are {}, {}.'.format(latitude, longitude))

The coordinates of Mumbai are 19.0759899, 72.8773928.


Installing folium for mapping.

In [15]:
pip install folium

  Using cached folium-0.12.1-py2.py3-none-any.whl (94 kB)
  Using cached branca-0.4.2-py3-none-any.whl (24 kB)
Note: you may need to restart the kernel to use updated packages.


In [16]:
import folium

In [22]:
map_Mumbai = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, location, area in zip(df['Latitude'], df['Longitude'], df['Location'], df['Area']):
    label = '{}, {}'.format(area, location)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Mumbai)  
    
map_Mumbai

Initializing the API credentials to use foursquare API to get nearby venues.

In [23]:
CLIENT_ID = 'SWJ0LEG2LUAIZKH5KCHUYP2P1MKSXJN5RZDWGFTRIYY1QS1S' 
CLIENT_SECRET = '0SL4GC0A2FDWL3YXFZDXECAHHBZNGRN11RYXGGIYOPKPXDPQ'
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SWJ0LEG2LUAIZKH5KCHUYP2P1MKSXJN5RZDWGFTRIYY1QS1S
CLIENT_SECRET:0SL4GC0A2FDWL3YXFZDXECAHHBZNGRN11RYXGGIYOPKPXDPQ


Creating a funciton that gets us the nearby venues of a given location.

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [26]:
venues_in_Mumbai = getNearbyVenues(df['Area'], df['Latitude'], df['Longitude'])

Amboli
Chakala, Andheri
D.N. Nagar
Four Bungalows
Lokhandwala
Marol
Sahar
Seven Bungalows
Versova
Mira Road
Bhayandar
Uttan
Bandstand Promenade
Kherwadi
Pali Hill
I.C. Colony
Gorai
Dahisar
Aarey Milk Colony
Bangur Nagar
Jogeshwari West
Juhu
Charkop
Poisar
Mahavir Nagar
Thakur village
Pali Naka
Khar Danda
Dindoshi
Sunder Nagar
Kalina
Naigaon
Nalasopara
Virar
Irla
Vile Parle
Bhandup
Amrut Nagar
Asalfa
Pant Nagar
Kanjurmarg
Nehru Nagar
Nahur
Chandivali
Hiranandani Gardens
Indian Institute of Technology Bombay campus
Vidyavihar
Vikhroli
Chembur
Deonar
Mankhurd
Mahul
Agripada
Altamount Road
Bhuleshwar
Breach Candy
Carmichael Road
Cavel
Churchgate
Cotton Green
Cuffe Parade
Cumbala Hill
Currey Road
Dhobitalao
Dongri
Kala Ghoda
Kemps Corner
Lower Parel
Mahalaxmi
Mahim
Malabar Hill
Marine Drive
Marine Lines
Mumbai Central
Nariman Point
Prabhadevi
Sion
Walkeshwar
Worli
C.G.S. colony
Dagdi Chawl
Navy Nagar
Hindu colony
Ballard Estate
Chira Bazaar
Fanas Wadi
Chor Bazaar
Matunga
Parel
Gowalia Tank


In [28]:
venues_in_Mumbai

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0           Amboli                 19.1293                  72.8434   
1           Amboli                 19.1293                  72.8434   
2           Amboli                 19.1293                  72.8434   
3           Amboli                 19.1293                  72.8434   
4           Amboli                 19.1293                  72.8434   
...            ...                     ...                      ...   
1142         Thane                 19.2000                  72.9700   
1143         Thane                 19.2000                  72.9700   
1144         Thane                 19.2000                  72.9700   
1145         Thane                 19.2000                  72.9700   
1146         Thane                 19.2000                  72.9700   

                       Venue       Venue Category  
0                  Cafe Arfa    Indian Restaurant  
1           5 Spice , Bandra   Chinese Restaurant  
2                     Subway       Sandwich Place  
3            Cafe Coffee Day          Coffee Shop  
4       Apple Service Centre          IT Services  
...                      ...                  ...  
1142   thane asead bus depot          Bus Station  
1143        Khopat Bus Stand          Bus Station  
1144         Vandana Talkies  Indie Movie Theater  
1145      Royal Challenge II    Indian Restaurant  
1146  Shree Krishna Icecream       Ice Cream Shop  

[1147 rows x 5 columns]

In [29]:
venues_in_Mumbai.groupby('Neighbourhood').head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0           Amboli                 19.1293                  72.8434   
1           Amboli                 19.1293                  72.8434   
2           Amboli                 19.1293                  72.8434   
3           Amboli                 19.1293                  72.8434   
4           Amboli                 19.1293                  72.8434   
...            ...                     ...                      ...   
1137         Thane                 19.2000                  72.9700   
1138         Thane                 19.2000                  72.9700   
1139         Thane                 19.2000                  72.9700   
1140         Thane                 19.2000                  72.9700   
1141         Thane                 19.2000                  72.9700   

                        Venue      Venue Category  
0                   Cafe Arfa   Indian Restaurant  
1            5 Spice , Bandra  Chinese Restaurant  
2                      Subway      Sandwich Place  
3             Cafe Coffee Day         Coffee Shop  
4        Apple Service Centre         IT Services  
...                       ...                 ...  
1137          Bombay Barbeque           BBQ Joint  
1138  Borivali Biryani Centre   Indian Restaurant  
1139                 FishLand  Seafood Restaurant  
1140                Fish Land  Seafood Restaurant  
1141        Harish Lunch Home  Seafood Restaurant  

[389 rows x 5 columns]

In [30]:
venues_in_Mumbai.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                               
ATM                       Mankhurd               19.050000   
Afghan Restaurant      Amrut Nagar               19.102077   
American Restaurant   Sunder Nagar               19.175000   
Amphitheater            Khar Danda               19.068598   
Antique Shop           Chor Bazaar               18.960321   
...                            ...                     ...   
Whisky Bar                   Parel               18.990000   
Wine Bar             Nariman Point               18.930000   
Wine Shop              Dava Bazaar               18.946882   
Women's Store              Matunga               19.130815   
Yoga Studio             Prabhadevi               19.016600   

                     Neighbourhood Longitude                          Venue  
Venue Category                                                               
ATM                                72.930000                  Axis Bank ATM  
Afghan Restaurant                  72.912835                        Zaffran  
American Restaurant                72.912835          Thank God It's Friday  
Amphitheater                       72.840042                    The Habitat  
Antique Shop                       72.827176  Chor Bazaar (Thieves' Market)  
...                                      ...                            ...  
Whisky Bar                         72.840000                    Best Punjab  
Wine Bar                           72.823000                      Opium Den  
Wine Shop                          72.831362                   Peekay Wines  
Women's Store                      72.844763                          Trios  
Yoga Studio                        72.829500     Moksh - The Wellness Place  

[163 rows x 4 columns]

So there are 163 different types of venue categories in Mumbai.

Now, we will filter those venue categories which have the term 'restaurant' in them.

In [31]:
resto_data = venues_in_Mumbai[venues_in_Mumbai['Venue Category'].str.contains('Restaurant', regex = False)]

In [32]:
resto_data

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0               Amboli               19.129300                72.843400   
1               Amboli               19.129300                72.843400   
5               Amboli               19.129300                72.843400   
8               Amboli               19.129300                72.843400   
12    Chakala, Andheri               19.111388                72.860833   
...                ...                     ...                      ...   
1138             Thane               19.200000                72.970000   
1139             Thane               19.200000                72.970000   
1140             Thane               19.200000                72.970000   
1141             Thane               19.200000                72.970000   
1145             Thane               19.200000                72.970000   

                        Venue        Venue Category  
0                   Cafe Arfa     Indian Restaurant  
1            5 Spice , Bandra    Chinese Restaurant  
5           Spices & Chillies      Asian Restaurant  
8                 Delhi Zaika      Halal Restaurant  
12                    Faaso's  Fast Food Restaurant  
...                       ...                   ...  
1138  Borivali Biryani Centre     Indian Restaurant  
1139                 FishLand    Seafood Restaurant  
1140                Fish Land    Seafood Restaurant  
1141        Harish Lunch Home    Seafood Restaurant  
1145       Royal Challenge II     Indian Restaurant  

[342 rows x 5 columns]

In [34]:
from sklearn.cluster import KMeans

Now we will use the k-means clustering to cluster the different restaurants only based on their coordinates. Later, when we will use clustering, it will be based on more than one features and then we will see how different both the maps look.

In [35]:
k=3
resto_clustering = resto_data.drop(['Neighbourhood','Venue','Venue Category'],1)
kmeans = KMeans(n_clusters = k,random_state=0).fit(resto_clustering)
kmeans.labels_
resto_data.insert(0, 'Cluster Labels', kmeans.labels_)

In [36]:
resto_data

Cluster Labels     Neighbourhood  Neighbourhood Latitude  \
0                  2            Amboli               19.129300   
1                  2            Amboli               19.129300   
5                  2            Amboli               19.129300   
8                  2            Amboli               19.129300   
12                 2  Chakala, Andheri               19.111388   
...              ...               ...                     ...   
1138               1             Thane               19.200000   
1139               1             Thane               19.200000   
1140               1             Thane               19.200000   
1141               1             Thane               19.200000   
1145               1             Thane               19.200000   

      Neighbourhood Longitude                    Venue        Venue Category  
0                   72.843400                Cafe Arfa     Indian Restaurant  
1                   72.843400         5 Spice , Bandra    Chinese Restaurant  
5                   72.843400        Spices & Chillies      Asian Restaurant  
8                   72.843400              Delhi Zaika      Halal Restaurant  
12                  72.860833                  Faaso's  Fast Food Restaurant  
...                       ...                      ...                   ...  
1138                72.970000  Borivali Biryani Centre     Indian Restaurant  
1139                72.970000                 FishLand    Seafood Restaurant  
1140                72.970000                Fish Land    Seafood Restaurant  
1141                72.970000        Harish Lunch Home    Seafood Restaurant  
1145                72.970000       Royal Challenge II     Indian Restaurant  

[342 rows x 6 columns]

In [44]:
# create map
map_clusters = folium.Map(location=[19.0759899,72.8773928],zoom_start=10)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, neighborhood, cluster in zip(resto_data['Neighbourhood Latitude'], resto_data['Neighbourhood Longitude'], resto_data['Neighbourhood'], resto_data['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [52]:
resto_data.drop(['Cluster Labels'],1)

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0               Amboli               19.129300                72.843400   
1               Amboli               19.129300                72.843400   
5               Amboli               19.129300                72.843400   
8               Amboli               19.129300                72.843400   
12    Chakala, Andheri               19.111388                72.860833   
...                ...                     ...                      ...   
1138             Thane               19.200000                72.970000   
1139             Thane               19.200000                72.970000   
1140             Thane               19.200000                72.970000   
1141             Thane               19.200000                72.970000   
1145             Thane               19.200000                72.970000   

                        Venue        Venue Category  
0                   Cafe Arfa     Indian Restaurant  
1            5 Spice , Bandra    Chinese Restaurant  
5           Spices & Chillies      Asian Restaurant  
8                 Delhi Zaika      Halal Restaurant  
12                    Faaso's  Fast Food Restaurant  
...                       ...                   ...  
1138  Borivali Biryani Centre     Indian Restaurant  
1139                 FishLand    Seafood Restaurant  
1140                Fish Land    Seafood Restaurant  
1141        Harish Lunch Home    Seafood Restaurant  
1145       Royal Challenge II     Indian Restaurant  

[342 rows x 5 columns]

One Hot Encoding the different venue categories.

In [53]:
resto_venue_cat = pd.get_dummies(resto_data[['Venue Category']], prefix="", prefix_sep="")
resto_venue_cat

Afghan Restaurant  American Restaurant  Asian Restaurant  \
0                     0                    0                 0   
1                     0                    0                 0   
5                     0                    0                 1   
8                     0                    0                 0   
12                    0                    0                 0   
...                 ...                  ...               ...   
1138                  0                    0                 0   
1139                  0                    0                 0   
1140                  0                    0                 0   
1141                  0                    0                 0   
1145                  0                    0                 0   

      Bengali Restaurant  Chinese Restaurant  Dim Sum Restaurant  \
0                      0                   0                   0   
1                      0                   1                   0   
5                      0                   0                   0   
8                      0                   0                   0   
12                     0                   0                   0   
...                  ...                 ...                 ...   
1138                   0                   0                   0   
1139                   0                   0                   0   
1140                   0                   0                   0   
1141                   0                   0                   0   
1145                   0                   0                   0   

      Falafel Restaurant  Fast Food Restaurant  French Restaurant  \
0                      0                     0                  0   
1                      0                     0                  0   
5                      0                     0                  0   
8                      0                     0                  0   
12                     0                     1                  0   
...                  ...                   ...                ...   
1138                   0                     0                  0   
1139                   0                     0                  0   
1140                   0                     0                  0   
1141                   0                     0                  0   
1145                   0                     0                  0   

      German Restaurant  ...  Mediterranean Restaurant  Mexican Restaurant  \
0                     0  ...                         0                   0   
1                     0  ...                         0                   0   
5                     0  ...                         0                   0   
8                     0  ...                         0                   0   
12                    0  ...                         0                   0   
...                 ...  ...                       ...                 ...   
1138                  0  ...                         0                   0   
1139                  0  ...                         0                   0   
1140                  0  ...                         0                   0   
1141                  0  ...                         0                   0   
1145                  0  ...                         0                   0   

      Middle Eastern Restaurant  Mughlai Restaurant  New American Restaurant  \
0                             0                   0                        0   
1                             0                   0                        0   
5                             0                   0                        0   
8                             0                   0                        0   
12                            0                   0                        0   
...                         ...                 ...                      ...   
1138                          0                   0                   

Adding the column 'Neighbourhood' to the one hot encoded table.

In [55]:
resto_venue_cat['Neighbourhood'] = resto_data['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [resto_venue_cat.columns[-1]] + list(resto_venue_cat.columns[:-1])
resto_venue_cat = resto_venue_cat[fixed_columns]

resto_venue_cat

Vegetarian / Vegan Restaurant     Neighbourhood  Afghan Restaurant  \
0                                 0            Amboli                  0   
1                                 0            Amboli                  0   
5                                 0            Amboli                  0   
8                                 0            Amboli                  0   
12                                0  Chakala, Andheri                  0   
...                             ...               ...                ...   
1138                              0             Thane                  0   
1139                              0             Thane                  0   
1140                              0             Thane                  0   
1141                              0             Thane                  0   
1145                              0             Thane                  0   

      American Restaurant  Asian Restaurant  Bengali Restaurant  \
0                       0                 0                   0   
1                       0                 0                   0   
5                       0                 1                   0   
8                       0                 0                   0   
12                      0                 0                   0   
...                   ...               ...                 ...   
1138                    0                 0                   0   
1139                    0                 0                   0   
1140                    0                 0                   0   
1141                    0                 0                   0   
1145                    0                 0                   0   

      Chinese Restaurant  Dim Sum Restaurant  Falafel Restaurant  \
0                      0                   0                   0   
1                      1                   0                   0   
5                      0                   0                   0   
8                      0                   0                   0   
12                     0                   0                   0   
...                  ...                 ...                 ...   
1138                   0                   0                   0   
1139                   0                   0                   0   
1140                   0                   0                   0   
1141                   0                   0                   0   
1145                   0                   0                   0   

      Fast Food Restaurant  ...  Maharashtrian Restaurant  \
0                        0  ...                         0   
1                        0  ...                         0   
5                        0  ...                         0   
8                        0  ...                         0   
12                       1  ...                         0   
...                    ...  ...                       ...   
1138                     0  ...                         0   
1139                     0  ...                         0   
1140                     0  ...                         0   
1141                     0  ...                         0   
1145                     0  ...                         0   

      Mediterranean Restaurant  Mexican Restaurant  Middle Eastern Restaurant  \
0                            0                   0                          0   
1                            0                   0                          0   
5                            0                   0                          0   
8                            0                   0                          0   
12                           0                   0                          0   
...                        ...                 ...                        ...   
1138                         0                   0                          0   
1139                         0                   0                          0   
1140     

Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [80]:
resto_grouped = resto_venue_cat.groupby('Neighbourhood').mean().reset_index()
resto_grouped.head()

Neighbourhood  Vegetarian / Vegan Restaurant  Afghan Restaurant  \
0             Agripada                            0.0           0.000000   
1       Altamount Road                            0.0           0.000000   
2               Amboli                            0.0           0.000000   
3          Amrut Nagar                            0.0           0.066667   
4  Bandstand Promenade                            0.0           0.000000   

   American Restaurant  Asian Restaurant  Bengali Restaurant  \
0             0.000000          0.000000                 0.0   
1             0.000000          0.000000                 0.0   
2             0.000000          0.250000                 0.0   
3             0.066667          0.066667                 0.0   
4             0.000000          0.000000                 0.0   

   Chinese Restaurant  Dim Sum Restaurant  Falafel Restaurant  \
0            0.000000                 0.0            0.000000   
1            0.000000                 0.0            0.000000   
2            0.250000                 0.0            0.000000   
3            0.066667                 0.0            0.066667   
4            0.250000                 0.0            0.000000   

   Fast Food Restaurant  ...  Maharashtrian Restaurant  \
0              0.000000  ...                       0.0   
1              0.000000  ...                       0.0   
2              0.000000  ...                       0.0   
3              0.133333  ...                       0.0   
4              0.250000  ...                       0.0   

   Mediterranean Restaurant  Mexican Restaurant  Middle Eastern Restaurant  \
0                  0.000000                 0.0                        0.0   
1                  0.000000                 0.0                        0.0   
2                  0.000000                 0.0                        0.0   
3                  0.066667                 0.0                        0.0   
4                  0.000000                 0.0                        0.0   

   Mughlai Restaurant  New American Restaurant  North Indian Restaurant  \
0                 0.0                      0.0                      0.0   
1                 0.0                      0.0                      0.0   
2                 0.0                      0.0                      0.0   
3                 0.0                      0.0                      0.0   
4                 0.0                      0.0                      0.0   

   Restaurant  Seafood Restaurant  South Indian Restaurant  
0    0.000000                 0.0                      0.0  
1    0.000000                 0.0                      0.0  
2    0.000000                 0.0                      0.0  
3    0.066667                 0.0                      0.0  
4    0.000000                 0.0                      0.0  

[5 rows x 29 columns]

Defining a function that returns the most common venus in each of the categories.

In [57]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [60]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = resto_grouped['Neighbourhood']

for ind in np.arange(resto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(resto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue    2nd Most Common Venue  \
0              Agripada     Indian Restaurant  South Indian Restaurant   
1        Altamount Road     Indian Restaurant  South Indian Restaurant   
2                Amboli      Halal Restaurant         Asian Restaurant   
3           Amrut Nagar     Indian Restaurant     Fast Food Restaurant   
4   Bandstand Promenade    Chinese Restaurant     Fast Food Restaurant   
..                  ...                   ...                      ...   
68                Uttan     Indian Restaurant  South Indian Restaurant   
69           Vidyavihar            Restaurant     Fast Food Restaurant   
70           Vile Parle     Indian Restaurant      Japanese Restaurant   
71           Walkeshwar  Fast Food Restaurant        Indian Restaurant   
72                Worli     Indian Restaurant     Fast Food Restaurant   

      3rd Most Common Venue    4th Most Common Venue    5th Most Common Venue  \
0          Greek Restaurant        Afghan Restaurant      American Restaurant   
1          Greek Restaurant        Afghan Restaurant      American Restaurant   
2        Chinese Restaurant        Indian Restaurant         Greek Restaurant   
3                Restaurant        Afghan Restaurant      American Restaurant   
4        Italian Restaurant        Indian Restaurant  South Indian Restaurant   
..                      ...                      ...                      ...   
68         Greek Restaurant        Afghan Restaurant      American Restaurant   
69  South Indian Restaurant         Greek Restaurant        Afghan Restaurant   
70       Seafood Restaurant  South Indian Restaurant          Goan Restaurant   
71  South Indian Restaurant         Greek Restaurant        Afghan Restaurant   
72       Seafood Restaurant  South Indian Restaurant         Greek Restaurant   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
0       Asian Restaurant    Bengali Restaurant        Chinese Restaurant   
1       Asian Restaurant    Bengali Restaurant        Chinese Restaurant   
2      Afghan Restaurant   American Restaurant        Bengali Restaurant   
3       Asian Restaurant    Chinese Restaurant  Mediterranean Restaurant   
4       Greek Restaurant     Afghan Restaurant       American Restaurant   
..                   ...                   ...                       ...   
68      Asian Restaurant    Bengali Restaurant        Chinese Restaurant   
69   American Restaurant      Asian Restaurant        Bengali Restaurant   
70     Afghan Restaurant   American Restaurant          Asian Restaurant   
71   American Restaurant      Asian Restaurant        Bengali Restaurant   
72     Afghan Restaurant   American Restaurant          Asian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0     Dim Sum Restaurant     Falafel Restaurant  
1     Dim Sum Restaurant     Falafel Restaurant  
2     Dim Sum Restaurant     Falafel Restaurant  
3     Falafel Restaurant     Italian Restaurant  
4       Asian Restaurant     Bengali Restaurant  
..                   ...                    ...  
68    Dim Sum Restaurant     Falafel Restaurant  
69    Chinese Restaurant     Dim Sum Restaurant  
70    Bengali Restaurant     Chinese Restaurant  
71    Chinese Restaurant     Dim Sum Restaurant  
72    Bengali Restaurant     Chinese Restaurant  

[73 rows x 11 columns]

In [61]:
k_num_clusters = 3

resto_grouped_clustering = resto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(resto_grouped_clustering)
kmeans

KMeans(n_clusters=3, random_state=0)

In [62]:
kmeans.labels_

array([0, 0, 1, 1, 1, 0, 1, 0, 1, 2, 0, 2, 2, 2, 0, 0, 0, 2, 1, 1, 0, 0,
       1, 1, 0, 2, 1, 2, 0, 1, 2, 1, 2, 1, 2, 0, 1, 1, 0, 2, 1, 1, 1, 1,
       0, 2, 1, 1, 1, 1, 1, 1, 1, 2, 2, 0, 2, 1, 2, 1, 0, 1, 1, 2, 0, 1,
       1, 1, 0, 2, 0, 1, 1], dtype=int32)

In [63]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [110]:
neighborhoods_venues_sorted.head()

Cluster Labels        Neighbourhood 1st Most Common Venue  \
0               0             Agripada     Indian Restaurant   
1               0       Altamount Road     Indian Restaurant   
2               1               Amboli      Halal Restaurant   
3               1          Amrut Nagar     Indian Restaurant   
4               1  Bandstand Promenade    Chinese Restaurant   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  South Indian Restaurant      Greek Restaurant     Afghan Restaurant   
1  South Indian Restaurant      Greek Restaurant     Afghan Restaurant   
2         Asian Restaurant    Chinese Restaurant     Indian Restaurant   
3     Fast Food Restaurant            Restaurant     Afghan Restaurant   
4     Fast Food Restaurant    Italian Restaurant     Indian Restaurant   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      American Restaurant      Asian Restaurant    Bengali Restaurant   
1      American Restaurant      Asian Restaurant    Bengali Restaurant   
2         Greek Restaurant     Afghan Restaurant   American Restaurant   
3      American Restaurant      Asian Restaurant    Chinese Restaurant   
4  South Indian Restaurant      Greek Restaurant     Afghan Restaurant   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Chinese Restaurant    Dim Sum Restaurant     Falafel Restaurant  
1        Chinese Restaurant    Dim Sum Restaurant     Falafel Restaurant  
2        Bengali Restaurant    Dim Sum Restaurant     Falafel Restaurant  
3  Mediterranean Restaurant    Falafel Restaurant     Italian Restaurant  
4       American Restaurant      Asian Restaurant     Bengali Restaurant

In order to plot these clustered neighbourhoods, we need to add thr coordinates of the neighbourhoods.

The coordinates are in the dataframe 'df'. 

In [64]:
df

Area                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

Renaming the 'area' column in df to join with the neighbourhood column of the 'neighborhoods_venues_sorted' table.

In [72]:
df = df.rename(columns = {'Area' : 'Neighbourhood'})

In [73]:
df

Neighbourhood                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

# Joining the two tables.

In [75]:
resto_merged = df.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

resto_merged

Neighbourhood                 Location   Latitude  Longitude  \
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400   
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833   
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373   
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210   
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270   
..               ...                      ...        ...        ...   
88             Parel             South Mumbai  18.990000  72.840000   
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703   
90       Dava Bazaar             South Mumbai  18.946882  72.831362   
91           Dharavi                   Mumbai  19.040208  72.850850   
92             Thane                   Mumbai  19.200000  72.970000   

    Cluster Labels          1st Most Common Venue    2nd Most Common Venue  \
0              1.0               Halal Restaurant         Asian Restaurant   
1              2.0                     Restaurant     Fast Food Restaurant   
2              0.0              Indian Restaurant  South Indian Restaurant   
3              2.0  Vegetarian / Vegan Restaurant               Restaurant   
4              0.0              Indian Restaurant  South Indian Restaurant   
..             ...                            ...                      ...   
88             1.0               Asian Restaurant        Indian Restaurant   
89             2.0                     Restaurant  South Indian Restaurant   
90             1.0              Indian Restaurant               Restaurant   
91             2.0           Fast Food Restaurant  South Indian Restaurant   
92             1.0             Seafood Restaurant        Indian Restaurant   

      3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0        Chinese Restaurant          Indian Restaurant      Greek Restaurant   
1          Asian Restaurant         Falafel Restaurant    Seafood Restaurant   
2          Greek Restaurant          Afghan Restaurant   American Restaurant   
3          Greek Restaurant          Afghan Restaurant   American Restaurant   
4          Greek Restaurant          Afghan Restaurant   American Restaurant   
..                      ...                        ...                   ...   
88  South Indian Restaurant           Greek Restaurant     Afghan Restaurant   
89         Greek Restaurant          Afghan Restaurant   American Restaurant   
90      American Restaurant  Middle Eastern Restaurant  Fast Food Restaurant   
91       Seafood Restaurant          Afghan Restaurant   American Restaurant   
92  South Indian Restaurant           Greek Restaurant     Afghan Restaurant   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Afghan Restaurant   American Restaurant    Bengali Restaurant   
1   South Indian Restaurant      Greek Restaurant     Afghan Restaurant   
2          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
3          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
4          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
..                      ...                   ...                   ...   
88      American Restaurant    Bengali Restaurant    Chinese Restaurant   
89         Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
90  South Indian Restaurant       Goan Restaurant     Afghan Restaurant   
91         Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
92      American Restaurant      Asian Restaurant    Bengali Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0     Dim Sum Restaurant     Falafel Restaurant  
1    American Restaurant     Bengali Restaurant  
2     Dim Sum Restaurant     Falafel Restaurant  
3     Dim Sum Restaurant     Falafel Restaurant  
4     Dim Sum Restaurant     Falafel Restaurant  
..                   ...                    ...  
88   

Dropping null values from the merged table.

In [76]:
resto_merged_nonan = resto_merged.dropna(subset=['Cluster Labels'])

In [92]:
resto_merged_nonan

Neighbourhood                 Location   Latitude  Longitude  \
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400   
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833   
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373   
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210   
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270   
..               ...                      ...        ...        ...   
88             Parel             South Mumbai  18.990000  72.840000   
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703   
90       Dava Bazaar             South Mumbai  18.946882  72.831362   
91           Dharavi                   Mumbai  19.040208  72.850850   
92             Thane                   Mumbai  19.200000  72.970000   

    Cluster Labels          1st Most Common Venue    2nd Most Common Venue  \
0              1.0               Halal Restaurant         Asian Restaurant   
1              2.0                     Restaurant     Fast Food Restaurant   
2              0.0              Indian Restaurant  South Indian Restaurant   
3              2.0  Vegetarian / Vegan Restaurant               Restaurant   
4              0.0              Indian Restaurant  South Indian Restaurant   
..             ...                            ...                      ...   
88             1.0               Asian Restaurant        Indian Restaurant   
89             2.0                     Restaurant  South Indian Restaurant   
90             1.0              Indian Restaurant               Restaurant   
91             2.0           Fast Food Restaurant  South Indian Restaurant   
92             1.0             Seafood Restaurant        Indian Restaurant   

      3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0        Chinese Restaurant          Indian Restaurant      Greek Restaurant   
1          Asian Restaurant         Falafel Restaurant    Seafood Restaurant   
2          Greek Restaurant          Afghan Restaurant   American Restaurant   
3          Greek Restaurant          Afghan Restaurant   American Restaurant   
4          Greek Restaurant          Afghan Restaurant   American Restaurant   
..                      ...                        ...                   ...   
88  South Indian Restaurant           Greek Restaurant     Afghan Restaurant   
89         Greek Restaurant          Afghan Restaurant   American Restaurant   
90      American Restaurant  Middle Eastern Restaurant  Fast Food Restaurant   
91       Seafood Restaurant          Afghan Restaurant   American Restaurant   
92  South Indian Restaurant           Greek Restaurant     Afghan Restaurant   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Afghan Restaurant   American Restaurant    Bengali Restaurant   
1   South Indian Restaurant      Greek Restaurant     Afghan Restaurant   
2          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
3          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
4          Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
..                      ...                   ...                   ...   
88      American Restaurant    Bengali Restaurant    Chinese Restaurant   
89         Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
90  South Indian Restaurant       Goan Restaurant     Afghan Restaurant   
91         Asian Restaurant    Bengali Restaurant    Chinese Restaurant   
92      American Restaurant      Asian Restaurant    Bengali Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0     Dim Sum Restaurant     Falafel Restaurant  
1    American Restaurant     Bengali Restaurant  
2     Dim Sum Restaurant     Falafel Restaurant  
3     Dim Sum Restaurant     Falafel Restaurant  
4     Dim Sum Restaurant     Falafel Restaurant  
..                   ...                    ...  
88   

Now, we will show the different clusters on the map.


In [77]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(resto_merged_nonan['Latitude'], resto_merged_nonan['Longitude'], resto_merged_nonan['Neighbourhood'], resto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

In [99]:
cluster1 = resto_merged_nonan.loc[resto_merged_nonan['Cluster Labels'] == 0, resto_merged_nonan.columns[[0] + [1]+ list(range(5, resto_merged_nonan.shape[1]))]]

In [100]:
cluster1.shape

(20, 12)

In [101]:
cluster1

Neighbourhood  \
2                                     D.N. Nagar   
4                                    Lokhandwala   
11                                         Uttan   
19                                  Bangur Nagar   
21                                          Juhu   
35                                    Vile Parle   
39                                    Pant Nagar   
42                                         Nahur   
45  Indian Institute of Technology Bombay campus   
48                                       Chembur   
49                                        Deonar   
52                                      Agripada   
53                                Altamount Road   
54                                    Bhuleshwar   
57                                         Cavel   
64                                        Dongri   
76                                          Sion   
80                                   Dagdi Chawl   
84                                  Chira Bazaar   
86                                   Chor Bazaar   

                          Location 1st Most Common Venue  \
2          Andheri,Western Suburbs     Indian Restaurant   
4          Andheri,Western Suburbs     Indian Restaurant   
11  Mira-Bhayandar,Western Suburbs     Indian Restaurant   
19        Goregaon,Western Suburbs     Indian Restaurant   
21                 Western Suburbs     Indian Restaurant   
35                 Western Suburbs     Indian Restaurant   
39       Ghatkopar,Eastern Suburbs     Indian Restaurant   
42          Mulund,Eastern Suburbs     Indian Restaurant   
45           Powai,Eastern Suburbs     Indian Restaurant   
48                 Harbour Suburbs     Indian Restaurant   
49         Govandi,Harbour Suburbs     Indian Restaurant   
52                    South Mumbai     Indian Restaurant   
53                    South Mumbai     Indian Restaurant   
54                    South Mumbai     Indian Restaurant   
57                    South Mumbai     Indian Restaurant   
64                    South Mumbai     Indian Restaurant   
76                    South Mumbai     Indian Restaurant   
80            Byculla,South Mumbai     Indian Restaurant   
84          Kalbadevi,South Mumbai     Indian Restaurant   
86        Kamathipura,South Mumbai     Indian Restaurant   

            2nd Most Common Venue    3rd Most Common Venue  \
2         South Indian Restaurant         Greek Restaurant   
4         South Indian Restaurant         Greek Restaurant   
11        South Indian Restaurant         Greek Restaurant   
19        South Indian Restaurant         Greek Restaurant   
21            Japanese Restaurant       Seafood Restaurant   
35            Japanese Restaurant       Seafood Restaurant   
39  Vegetarian / Vegan Restaurant               Restaurant   
42                     Restaurant  South Indian Restaurant   
45        South Indian Restaurant         Greek Restaurant   
48           Fast Food Restaurant  South Indian Restaurant   
49        South Indian Restaurant         Greek Restaurant   
52        South Indian Restaurant         Greek Restaurant   
53        South Indian Restaurant         Greek Restaurant   
54                     Restaurant      American Restaurant   
57             Chinese Restaurant  South Indian Restaurant   
64        South Indian Restaurant         Greek Restaurant   
76  Vegetarian / Vegan Restaurant         Greek Restaurant   
80        South Indian Restaurant         Greek Restaurant   
84        South Indian Restaurant         Greek Restaurant   
86                     Restaurant  South Indian Restaurant   

      4th Most Common Venue    5th Most Common Venue 6th Most Common Venue  \
2         Afghan Restaurant      American Restaurant      Asian Restaurant   
4         Afghan Restaurant      American Restaurant      Asian Restaurant   
11        Afghan Restaurant      American Restaurant      Asian Restaurant   
19        Afghan Restaurant      American Restaurant      Asian Re

In [102]:
cluster2 = resto_merged_nonan.loc[resto_merged_nonan['Cluster Labels'] == 1, resto_merged_nonan.columns[[0] + [1] + list(range(5, resto_merged_nonan.shape[1]))]]

In [103]:
cluster2.shape

(35, 12)

In [105]:
cluster2

Neighbourhood                         Location  \
0                Amboli          Andheri,Western Suburbs   
5                 Marol          Andheri,Western Suburbs   
12  Bandstand Promenade           Bandra,Western Suburbs   
13             Kherwadi           Bandra,Western Suburbs   
16                Gorai  Borivali (West),Western Suburbs   
17              Dahisar                  Western Suburbs   
20      Jogeshwari West                  Western Suburbs   
24        Mahavir Nagar   Kandivali West,Western Suburbs   
25       Thakur village   Kandivali East,Western Suburbs   
26            Pali Naka             Khar,Western Suburbs   
27           Khar Danda             Khar,Western Suburbs   
29         Sunder Nagar            Malad,Western Suburbs   
30               Kalina       Sanctacruz,Western Suburbs   
34                 Irla       Vile Parle,Western Suburbs   
36              Bhandup                  Eastern Suburbs   
37          Amrut Nagar        Ghatkopar,Eastern Suburbs   
44  Hiranandani Gardens            Powai,Eastern Suburbs   
55         Breach Candy                     South Mumbai   
61         Cumbala Hill                     South Mumbai   
62          Currey Road                     South Mumbai   
63           Dhobitalao                     South Mumbai   
66         Kemps Corner                     South Mumbai   
69                Mahim                     South Mumbai   
70         Malabar Hill                     South Mumbai   
71         Marine Drive                     South Mumbai   
72         Marine Lines                     South Mumbai   
74        Nariman Point                     South Mumbai   
75           Prabhadevi                     South Mumbai   
77           Walkeshwar                     South Mumbai   
78                Worli                     South Mumbai   
85           Fanas Wadi           Kalbadevi,South Mumbai   
87              Matunga                     South Mumbai   
88                Parel                     South Mumbai   
90          Dava Bazaar                     South Mumbai   
92                Thane                           Mumbai   

            1st Most Common Venue          2nd Most Common Venue  \
0                Halal Restaurant               Asian Restaurant   
5               Indian Restaurant                     Restaurant   
12             Chinese Restaurant           Fast Food Restaurant   
13              Indian Restaurant  Vegetarian / Vegan Restaurant   
16              Indian Restaurant             Seafood Restaurant   
17                     Restaurant              Indian Restaurant   
20           Fast Food Restaurant              Indian Restaurant   
24              Indian Restaurant           Fast Food Restaurant   
25              Indian Restaurant                     Restaurant   
26              Indian Restaurant             Seafood Restaurant   
27              Indian Restaurant                     Restaurant   
29              Indian Restaurant             Chinese Restaurant   
30              Indian Restaurant      Middle Eastern Restaurant   
34           Fast Food Restaurant              Indian Restaurant   
36  Vegetarian / Vegan Restaurant               Asian Restaurant   
37              Indian Restaurant           Fast Food Restaurant   
44              Indian Restaurant                     Restaurant   
55            Japanese Restaurant             Italian Restaurant   
61            Japanese Restaurant             Italian Restaurant   
62       Maharashtrian Restaurant              Indian Restaurant   
63              Indian Restaurant             Chinese Restaurant   
66           Fast Food Restaurant            Japanese Restaurant   
69              Indian Restaurant             Chinese Restaurant   
70           Fast Food Restaurant              Indian Restaurant   
71             Chinese Restaurant             Italian Restaurant   
72              Indian Restaurant             Chinese Restaurant   
74           

In [107]:
cluster3 = resto_merged_nonan.loc[resto_merged_nonan['Cluster Labels'] == 2, resto_merged_nonan.columns[[0] + [1] + list(range(5, resto_merged_nonan.shape[1]))]]

In [108]:
cluster3.shape

(18, 12)

In [109]:
cluster3

Neighbourhood                         Location  \
1   Chakala, Andheri                  Western Suburbs   
3     Four Bungalows          Andheri,Western Suburbs   
7    Seven Bungalows          Andheri,Western Suburbs   
9          Mira Road   Mira-Bhayandar,Western Suburbs   
14         Pali Hill           Bandra,Western Suburbs   
15       I.C. Colony  Borivali (West),Western Suburbs   
22           Charkop   Kandivali West,Western Suburbs   
28          Dindoshi            Malad,Western Suburbs   
32        Nalasopara            Vasai,Western Suburbs   
43        Chandivali            Powai,Eastern Suburbs   
46        Vidyavihar                  Eastern Suburbs   
56   Carmichael Road                     South Mumbai   
58        Churchgate                     South Mumbai   
65        Kala Ghoda                     South Mumbai   
67       Lower Parel                     South Mumbai   
73    Mumbai Central                     South Mumbai   
89      Gowalia Tank              Tardeo,South Mumbai   
91           Dharavi                           Mumbai   

            1st Most Common Venue          2nd Most Common Venue  \
1                      Restaurant           Fast Food Restaurant   
3   Vegetarian / Vegan Restaurant                     Restaurant   
7              Chinese Restaurant             Seafood Restaurant   
9   Vegetarian / Vegan Restaurant             Chinese Restaurant   
14           Fast Food Restaurant             Italian Restaurant   
15             Chinese Restaurant           Fast Food Restaurant   
22             Chinese Restaurant        South Indian Restaurant   
28                     Restaurant           Fast Food Restaurant   
32             Seafood Restaurant        South Indian Restaurant   
43      Indian Chinese Restaurant        South Indian Restaurant   
46                     Restaurant           Fast Food Restaurant   
56             Chinese Restaurant           Fast Food Restaurant   
58                     Restaurant             Italian Restaurant   
65             Seafood Restaurant               Asian Restaurant   
67             Italian Restaurant  Vegetarian / Vegan Restaurant   
73           Fast Food Restaurant  Vegetarian / Vegan Restaurant   
89                     Restaurant        South Indian Restaurant   
91           Fast Food Restaurant        South Indian Restaurant   

      3rd Most Common Venue      4th Most Common Venue  \
1          Asian Restaurant         Falafel Restaurant   
3          Greek Restaurant          Afghan Restaurant   
7   South Indian Restaurant    North Indian Restaurant   
9        Seafood Restaurant          Afghan Restaurant   
14  New American Restaurant  Middle Eastern Restaurant   
15        Indian Restaurant    South Indian Restaurant   
22       Seafood Restaurant          Afghan Restaurant   
28  South Indian Restaurant           Greek Restaurant   
32        Afghan Restaurant        American Restaurant   
43         Greek Restaurant          Afghan Restaurant   
46  South Indian Restaurant           Greek Restaurant   
56       Bengali Restaurant         Italian Restaurant   
58         Asian Restaurant   Mediterranean Restaurant   
65       Mughlai Restaurant         Chinese Restaurant   
67         Asian Restaurant         Chinese Restaurant   
73               Restaurant         Chinese Restaurant   
89         Greek Restaurant          Afghan Restaurant   
91       Seafood Restaurant          Afghan Restaurant   

       5th Most Common Venue    6th Most Common Venue  \
1         Seafood Restaurant  South Indian Restaurant   
3        American Restaurant         Asian Restaurant   
7         Dim Sum Restaurant        Indian Restaurant   
9        American Restaurant         Asian Restaurant   
14         Indian Restaurant         Greek Restaurant   
15          Greek Restaurant        Afghan Restaurant   
22       American Restaurant         Asian Restaurant   
28         Afghan Restaurant      American Restaurant   
32          As

## Observations:

As we can observe, cluster1 has a dominance of Indian restaurants with it being the most common venue followed by South Indian restaurants. Hence, anyone planning to open an Indian or South Indian restaurant in these regions will not be profitable as there would be intense competition already.
Cluster2 has a mix of different restaurants so any restaurant you open in those regions will result in good enough competition, but it would not be the worst decision to open a restaurant here.

Cluster3, as we can see, is dominated by resturants that are not Indian or south Indian. Hence, it would be profitable if someone opens an Indian restuarant here or even South Indian. The neighbourhood in this cluster has enough of Chinese and italian and seafood restaurant so it would not be advisable to invest in those kinds of restaurants.

In a nutshell, we can say the following overall:
    i) Neighbourhoods in cluster1 - Any restaurant but Indian or South Indian.
    ii) Neighbourhoods in cluster2 - Every restaurant will have competition already. Still, no Indian restaurant as they are still in plenty here.
    iii) Neighbourhoods in cluster3 - Indian restaurants are not much here. So, they will have good odds of making a profit here. Chinese restauras are already there in abundance so it would be good to avoid establishing another one here.

## Acknowledgement

Conclusion:
I have completed the process of identifying the business problems, mentioning the data required, preprocessing the data, visualizing the results, performing machine learning by k-means, clustering the data into 3 clusters based on their frequency similarities, and then reaching to a definitive solution to the business problems. the aim of the project is to provide recommendations to the relevant investors regarding the best kinds of restaurants that can be opened in a given location, or to find the best location to open a definite kind of a restaurant . 